pandas/_libs/groupby.pyx

```diff
@@ -398,11 +398,12 @@ def group_cumsum(
         for i in range(N):
             lab = labels[i]
 
-            if lab < 0:
+            if uses_mask and lab < 0:
```

Comment:
If the solution for the bug fix mentioned in #58984 is the only one available, I believe some regression may be unavoidable.